consensus/src/dag/reliable_broadcast.rs

```diff
@@ -184,3 +186,53 @@ impl RpcHandler for NodeBroadcastHandler {
         }
     }
 }
+
+#[derive(Debug, ThisError)]
+pub enum CertifiedNodeHandleError {
+    #[error("node already exists")]
+    NodeExists,
+    #[error("missing parents")]
+    MissingParents,
+}
+
+pub struct CertifiedNodeHandler {
+    dag: Arc<RwLock<Dag>>,
+}
+
+impl CertifiedNodeHandler {
+    pub fn new(dag: Arc<RwLock<Dag>>) -> Self {
+        Self { dag }
+    }
+}
+
+impl RpcHandler for CertifiedNodeHandler {
+    type Request = CertifiedNode;
+    type Response = CertifiedAck;
+
+    fn process(&mut self, node: Self::Request) -> anyhow::Result<Self::Response> {
+        {
+            let dag_reader = self.dag.read();
+            if dag_reader.exists(&node.digest()) {
+                bail!(CertifiedNodeHandleError::NodeExists);
+            }
+
+            if node.metadata().round() > 0
+                && (node.parents().len() == 0
+                    || !dag_reader.all_exists(
+                        node.parents()
+                            .iter()
+                            .map(|parent| parent.metadata().digest()),
+                    ))
+            {
+                // TODO(ibalajiarun): implement fetching logic.
+                bail!(CertifiedNodeHandleError::MissingParents);
```

Comment:
So we do not return ack in this case even though the validator got the message?
The other side will keep sending. 
Are we mixing reliable broadcast implementation with fetching?

Can we store the certificate internally and return ack instead?

Reply:
I don't think we should send back the ack without having all the parents if we want to guarantee that certificates mean nodes have all history.

I imagine if we plug in the dag_fetcher, it'd have a callback to send the ack if all missing parents are fetched

Reply:
Currently, Balaji is sending back the error. So the sender can choose to treat it as an ack.